user-manual/src/config/kdl.md

```diff
@@ -255,6 +264,121 @@ Filters at this stage are the earliest. Currently supported filters:
+is created, where one "token" is required for each request.
+
+The bucket for a rule starts with a configurable `tokens-per-bucket` number. When a request arrives,
+it attempts to take one token from the bucket. If one is available, it is served immediately. Otherwise,
+the request waits in a first-in, first-out order for a token to become available.
+
+The bucket is refilled at a configurable rate, specified by `refill-rate-ms`, and adds a configurable
+number of tokens specified by `refill-qty`. The number of tokens in the bucket will never exceed the
+initial `tokens-per-bucket` number.
+
+Once a refill occurs, requests may become ready if a token becomes available.
+
+[Leaky Bucket]: https://en.wikipedia.org/wiki/Leaky_bucket
+
+##### How many buckets?
+
+Some rules require many buckets. For example, rules based on the source IP address will create a bucket
+for each unique IP address of downstream users.
+
+However, each of these buckets require space to contain the metadata, and to avoid unbounded growth,
+we allow for a configurable `max-buckets` number, which serves to influence the total memory required
+for storing buckets. This uses an [Adaptive Replacement Cache]
+to allow for concurrent access to these buckets, as well as the ability to automatically buckets that
+are not actively being used (somewhat similar to an LRU or "Least Recently Used" cache).
+
+[Adaptive Replacement Cache]: https://docs.rs/concread/latest/concread/arcache/index.html
+
+There is a trade off here: The larger `max-buckets` is, the longer that River can "remember" a bucket
+for a given factor, such as specific IP addresses. However, it also requires more resident memory to
+retain this information.
+
+If `max-buckets` is set too low, then buckets will be "evicted" from the cache, meaning that subsequent
+requests matching that bucket will require the creation of a new bucket (with a full set of tokens),
+potentially defeating the objective of accurate rate limiting.
+
+##### Gotta claim 'em all
+
+When multiple rules apply to a single request, for example rules based on both source IP address,
+and the URI path, then a request must claim ALL applicable tokens before proceeding. If a given IP
+address is making it's first request, but to a URI that that has an empty bucket, it will immediately
+obtain the IP address token, but be forced to wait until the URI token has been claimed
+
+##### Kinds of Rules
+
+Currently two kinds of rules are supported:
+
+* `kind="source-ip"` - this tracks the IP address of the requestor. A unique bucket will be created for
+  the IPv4 or IPv6 address of the requestor.
+* `kind="uri" pattern="REGEX"` - This tracks the URI path of the request, such as `static/images/example.jpg`
+    * If the request's URI path matches the provided `REGEX`, the full URI path will be assigned to a given
+      bucket
+    * For example, if the regex `static/.*` was provided:
```

Comment:
Q. What if there are multiple rules that match a given request URL? Is there an ordering/hierarchy of rules, or would the request match every rule with a matching regex?

Reply:
Right now, it is specified as "tokens must be acquired for all matching rules". So if you had two URI rules like:

* `static/.*`
* `.*\.jpg`

And requested `static/images/example.jpg`, you would match both rules, and need to get tokens for each rule.

I chose this behavior because it matches what NGINX does:

> You can include multiple limit_req directives in a single location. All limits that match a given request are applied, meaning the most restrictive one is used. For example, if more than one directive imposes a delay, the longest delay is used. Similarly, requests are rejected if that is the effect of any directive, even if other directives allow them through.

I'm open to other behaviors, but I don't have the context to know what other behavior  people could prefer here!

Reply:
I'm going to resolve this, but @branlwyd please do open an issue if there's another approach you'd prefer or would like to discuss in general!

Reply:
Fair enough. I was thinking of a situation where you might want a low rate limit for most paths (e.g. `.*`), but a high rate limit for a few paths (e.g. `/static/.*`). In that case, you might want some way to only match the higher-limit rule, via hierarchy or priority ordering of rules. However, I _think_ such exclusions could also be implemented via a sufficiently-advanced-regex (e.g. in the previous example, the low-rate-limit rule would use a regex which matches everything but `/static/.*`), so this definitely is a nice-to-have.